diff -Naurp a/misc/paths.py.in b/misc/paths.py.in
--- a/misc/paths.py.in	2018-07-25 00:01:28.000000000 +0200
+++ b/misc/paths.py.in	2019-08-03 11:56:57.032059619 +0200
@@ -26,6 +26,7 @@
 
 import os
 import sys
+import site
 
 # some scripts expect this attribute to be in this module
 prefix = '@prefix@'
@@ -58,9 +59,9 @@ sys.path.insert(0, prefix)
 sys.path.insert(0, os.path.join(prefix, 'pythonlib'))
 
 # Include Python's site-packages directory.
-sitedir = os.path.join(sys.prefix, 'lib', 'python'+sys.version[:3],
-                       'site-packages')
-sys.path.append(sitedir)
+sitedirs = site.getsitepackages()
+for sitedir in sitedirs:
+    sys.path.append(sitedir)
 
 # Include Python's dist-packages directory.
 distdir = os.path.join(sys.prefix, 'lib', 'python'+sys.version[:3],
